### PR TITLE
Implement `CollectReq` boundary check

### PR DESCRIPTION
**Pull request by nakatsuka-y**
This PR implements a mechanism to handle `CollectReq` properly (#45).

A malicious Collector may request aggregation results over overlapping intervals.
This will  allow the malicious Collector to obtain aggregation results that are in batches smaller than `min_batch_interval`.

The Leader holds a binary tree for each Collector (distinguished via bearer token) per task.
The nodes in the binary tree hold the Interval requested by the Collector for this task in the past.
When the Leader receives a CollectReq from a Collector, it searches the binary tree for an overlapping node and if so, will reject the request.
If the new CollectReq does not overlap with the past intervals, then the Leader stores the new Interval into the tree and proceed to store the collect Id and request.

Unit test cases for the binary tree is written.
Another test case where the Collector tries to send a collect request with an overlapping interval is also implemented.

### CODE REVIEW

**Review by cjpatton — COMMENTED** [review]
Based on a quick look and reading the top comment this seems like a great approach! I suspect that O(log n) is pretty close to optimal runtime for this check. A few comments before diving in:
1. Enforcement per collector (i.e., bearer token) is not the right level of abstraction. The collector authentication mechanism is one piece of the protocol that is likely to change, so we'll need to decouple collect request enforcement from bearer tokens. I think enforcement per task would be better.
2. This mechanism is going to have to be implemented by both MockAggregator and DaphneWorkerConfig. Because we have to implement it twice and get good test coverage for both: (a) we need to maximize the amount of code shared by both; and (b) the pieces that must be implemented separately by both should be as simple as possible.
3. Bare in mind that it's more important that DaphneWorkerConfig is efficient than MockAggreegator.

File: daphne_worker/src/durable/aggregate_store.rs [hunk 1]
```diff
@@ -52,9 +61,21 @@ impl DurableObject for AggregateStore {
             }
 
             (DURABLE_AGGREGATE_STORE_GET, Method::Post) => {
-                let agg_share: DapAggregateShare =
-                    state_get_or_default(&self.state, "agg_share").await?;
-                Response::from_json(&agg_share)
+                let collected: bool = state_get_or_default(&self.state, "collected").await?;
+                if !collected {
+                    let agg_share: DapAggregateShare =
+                        state_get_or_default(&self.state, "agg_share").await?;
```
Comment (cjpatton) [inline daphne_worker/src/durable/aggregate_store.rs:67]: It's posible to get `collected` and `agg_share` concurrently, which would save on latency:
```rust
let agg_share_future = state_get_or_default(&self.state, "agg_share");
let collected: bool = state_get_or_default(&self.state, "collected").await?;
if !collected {
    let agg_sahre: DapAggregateShare = agg_share_future.await?;
}
```

**Review by cjpatton — CHANGES_REQUESTED** [review]
Looking good! I've only looked at Daphne-Worker so far.

File: daphne/src/auth.rs [hunk 2]
```diff
@@ -17,7 +17,7 @@ use serde::{Deserialize, Serialize};
 use std::io::Cursor;
 
 /// A bearer token used for authorizing DAP requests as specified in draft-ietf-ppm-dap-01.
-#[derive(Clone, Debug, Deserialize, Serialize)]
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Hash, Serialize)]
```
[the remark this answers is not in the document]
Comment (nakatsuka-y) [reply daphne/src/auth.rs]: This was from a past change. Thanks for pointing it out, I've removed it.

File: daphne/src/messages.rs [hunk 3]
```diff
@@ -288,6 +288,7 @@ pub enum TransitionFailure {
     HpkeUnknownConfigId = 3,
     HpkeDecryptError = 4,
     VdafPrepError = 5,
+    InvalidBatchInterval = 6,
```
[the remark this answers is not in the document]
Comment (nakatsuka-y) [reply daphne/src/messages.rs]: I've removed this and replaced with `DapError::PeerError`.

File: daphne_worker/src/dap.rs [hunk 4]
```diff
@@ -192,6 +193,7 @@ impl<D> DapAggregator<BearerToken> for DaphneWorkerConfig<D> {
         let namespace = self.durable_object("DAP_AGGREGATE_STORE")?;
         let task_id_base64url = task_id.to_base64url();
         let mut agg_share = DapAggregateShare::default();
+        // TODO(nakatsuka-y) Restrict the batch_interval or the following iteration will be expensive.
```
[the remark this answers is not in the document]
Comment (nakatsuka-y) [reply daphne_worker/src/dap.rs:197]: I've added changes to incorporate this check.

File: daphne_worker/src/dap.rs [hunk 5]
```diff
@@ -200,13 +202,13 @@ impl<D> DapAggregator<BearerToken> for DaphneWorkerConfig<D> {
...
-                    .await?
-                    .json()
-                    .await?;
+            let mut resp = durable_post!(stub, DURABLE_AGGREGATE_STORE_GET, &()).await?;
 
-            agg_share.merge(agg_share_delta)?;
+            // If this agg share has been collected before, return BatchCollected error.
+            match resp.json().await? {
+                AggregateStoreResult::Ok(agg_share_delta) => agg_share.merge(agg_share_delta)?,
+                AggregateStoreResult::Err(t) => return Err(DapError::Transition(t)),
```
[the remark this answers is not in the document]
Comment (nakatsuka-y) [reply daphne_worker/src/dap.rs]: Changed this to `DapError::PeerError`

File: daphne_worker/src/durable/aggregate_store.rs [hunk 6]
```diff
@@ -14,6 +15,14 @@ pub(crate) fn durable_agg_store_name(task_id_base64url: &str, window: u64) -> St
 
 pub(crate) const DURABLE_AGGREGATE_STORE_GET: &str = "/internal/do/aggregate_store/get";
 pub(crate) const DURABLE_AGGREGATE_STORE_MERGE: &str = "/internal/do/aggregate_store/merge";
+pub(crate) const DURABLE_AGGREGATE_STORE_MARK_COLLECTED: &str =
+    "/internal/do/aggregate_store/mark_collected";
+
+#[derive(Deserialize, Serialize)]
+pub(crate) enum AggregateStoreResult {
+    Ok(DapAggregateShare),
+    Err(TransitionFailure),
```
[the remark this answers is not in the document]
Comment (nakatsuka-y) [reply daphne_worker/src/durable/aggregate_store.rs]: Ditto :)

File: daphne_worker/src/durable/aggregate_store.rs [hunk 7]
```diff
@@ -52,9 +61,21 @@ impl DurableObject for AggregateStore {
             }
 
             (DURABLE_AGGREGATE_STORE_GET, Method::Post) => {
-                let agg_share: DapAggregateShare =
-                    state_get_or_default(&self.state, "agg_share").await?;
-                Response::from_json(&agg_share)
+                let collected: bool = state_get_or_default(&self.state, "collected").await?;
```
[the remark this answers is not in the document]
Comment (nakatsuka-y) [reply daphne_worker/src/durable/aggregate_store.rs:69]: Added note :)

File: daphne_worker/src/durable/aggregate_store.rs (hunk 1 above)
Comment (nakatsuka-y) [reply daphne_worker/src/durable/aggregate_store.rs]: Thanks for the suggestion!

File: daphne/src/lib.rs [hunk 8]
```diff
@@ -48,6 +48,10 @@ pub enum DapError {
     #[error("fatal error: {0}")]
     Fatal(String),
 
+    /// Error caused due to messages from a peer.
+    #[error("peer error: {0}")]
+    PeerError(String),
```
Comment (cjpatton) [inline daphne/src/lib.rs:53]: If you do this instead, you can get rid of the hard-coded stuff. By the way, you're going to need to add a new abort type: See the changes in pr #297 on the DAP spec.
```suggestion
    PeerError(DapAbort),
```

File: daphne/src/lib.rs [hunk 9]
```diff
@@ -227,13 +238,19 @@ pub struct ProblemDetails {
...
+// TODO spec: `max_window_size`, `min_duration_start`, and `max_duration_end`
+//            should be reflected to the DAP spec.
 #[derive(Deserialize, Serialize)]
 #[serde(try_from = "ShadowDapTaskConfig")]
 pub struct DapTaskConfig {
     pub leader_url: Url,
     pub helper_url: Url,
     pub min_batch_duration: u64, // seconds
     pub min_batch_size: u64,     // number of reports
+    pub max_window_size: u64,    // number of acceptable batches of reports
```
Comment (cjpatton) [inline daphne/src/lib.rs:251]: Let's do this instead
```suggestion
    /// Number of acceptable batches of reports.
    pub max_window_size: u64,
```

Also, this comment is not very descriptive. It's especially important to be clear here since this isn't in the spec.

File: daphne/src/roles_test.rs [hunk 10]
```diff
@@ -144,13 +144,12 @@ impl MockAggregator {
 
         // Construct report.
         let vdaf_config: &VdafConfig = &VdafConfig::Prio3(Prio3Config::Count);
+        let now = SystemTime::now()
```
Comment (cjpatton) [inline daphne/src/roles_test.rs:147]: Can we avoid using the system time for this test case? See #54.

File: daphne_worker/src/durable/aggregate_store.rs [hunk 11]
```diff
@@ -61,14 +62,16 @@ impl DurableObject for AggregateStore {
...
+                    state_get_or_default(&self.state, "agg_share").await?;
                 let collected: bool = state_get_or_default(&self.state, "collected").await?;
                 if !collected {
-                    let agg_share: DapAggregateShare =
-                        state_get_or_default(&self.state, "agg_share").await?;
                     Response::from_json(&AggregateStoreResult::Ok(agg_share))
                 } else {
                     Response::from_json(&AggregateStoreResult::Err(
-                        TransitionFailure::InvalidBatchInterval,
+                        PEER_ERR_BATCH_OVERLAP.to_string(),
```
Comment (cjpatton) [inline daphne_worker/src/durable/aggregate_store.rs:74]: This should be a `DapAbort` variant rather than a string.

**Review by cjpatton — CHANGES_REQUESTED** [review]

**Comment by cjpatton** [comment]
High level comment: I think we should move the new max_window_size, ... parameters out of VdafConfig. They should be picked globally per deployment, not per task.

File: daphne/src/lib.rs (hunk 8 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs]: We are now using DapAbort instead of String.

File: daphne/src/lib.rs (hunk 9 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs]: Added a more descriptive comment and moved the three variables out of task_config and into their own config struct.

File: daphne_worker/src/durable/aggregate_store.rs (hunk 11 above)
Comment (nakatsuka-y) [reply daphne_worker/src/durable/aggregate_store.rs]: Since we cannot use DapAbort (we cannot attach `Serialize`, `Deserialize` which is necessary to use it in `from_json`), we defined a new enum `ErrBatchOverlap`.

File: daphne/src/lib.rs [hunk 12]
```diff
@@ -225,6 +236,45 @@ pub struct ProblemDetails {
     pub detail: Option<String>,
 }
 
+/// Global DAP parameters common across tasks.
+//
+// TODO spec: `max_window_size`, `min_duration_start`, and `max_duration_end`
+//            should be reflected to the DAP spec.
+#[derive(Deserialize, Serialize)]
+#[serde(try_from = "ShadowDapGlobConfig")]
```
Comment (cjpatton) [inline daphne/src/lib.rs:244]: This shadow struct isn't needed because we don't have any fields that require custom serialization.
```suggestion
```

File: daphne/src/lib.rs [hunk 13]
```diff
@@ -225,6 +236,45 @@ pub struct ProblemDetails {
...
+pub struct DapGlobConfig {
+    /// Maximum window size permitted in CollectReq.
+    /// Prevents Collectors from requesting wide range or reports.
+    pub max_window_size: u64,
```
Comment (cjpatton) [inline daphne/src/lib.rs:248]: I think a better name would be "max_batch_duration", as it aligns better with "min_batch_duration".

File: daphne/src/lib.rs [hunk 14]
```diff
@@ -225,6 +236,45 @@ pub struct ProblemDetails {
...
+
+    /// Lower bound of an acceptable batch window in CollectReq.
+    /// Batch intervals must start after `min_duration_start`.
+    pub min_duration_start: u64,
```
Comment (cjpatton) [inline daphne/src/lib.rs:252]:
```suggestion
    pub min_batch_interval_start: u64,
```

File: daphne/src/lib.rs [hunk 15]
```diff
@@ -225,6 +236,45 @@ pub struct ProblemDetails {
...
+
+    /// Higher bound of an acceptable batch window in CollectReq.
+    /// Batch intervals must end before `max_duration_end`.
+    pub max_duration_end: u64,
```
Comment (cjpatton) [inline daphne/src/lib.rs:256]: Is this a duration (i.e., a number of seconds) or a timestamp (i.e., number of seconds after UNIX epoch)?

File: daphne/src/lib.rs [hunk 16]
```diff
@@ -225,6 +236,45 @@ pub struct ProblemDetails {
...
+}
+
+#[derive(Deserialize, Serialize)]
+struct ShadowDapGlobConfig {
```
Comment (cjpatton) [inline daphne/src/lib.rs:260]: "Glob" is a word in English that means "a lump of a semiliquid substance" :)
```suggestion
struct ShadowDapGlobalConfig {
```

File: daphne_worker/src/config.rs [hunk 17]
```diff
@@ -60,6 +61,10 @@ pub(crate) struct DaphneWorkerConfig<D> {
 impl<D> DaphneWorkerConfig<D> {
     /// Fetch DAP parameters from environment variables.
     pub(crate) fn from_worker_context(ctx: RouteContext<D>) -> Result<Self> {
+        let glob_config: DapGlobConfig =
```
Comment (cjpatton) [inline daphne_worker/src/config.rs:64]:
```suggestion
        let global_config: DapGlobalConfig =
```

File: daphne/src/lib.rs [hunk 18]
```diff
@@ -48,6 +48,10 @@ pub enum DapError {
     #[error("fatal error: {0}")]
     Fatal(String),
 
+    /// Error caused due to messages from a peer.
+    #[error("peer error: {0}")]
+    PeerError(DapAbort),
```
Comment (cjpatton) [inline daphne/src/lib.rs:53]:
```suggestion
    /// Error triggered by peer, resulting in an abort.
    #[error("abort: {0}")]
    Abort(DapAbort),
```

File: daphne_worker/src/dap.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cjpatton) [reply daphne_worker/src/dap.rs:179]: Great, in that case, pleases remove the comment.

File: daphne_worker/src/lib.rs [hunk 19]
```diff
@@ -14,6 +14,12 @@ use prio::codec::{Decode, Encode};
 use serde::{Deserialize, Serialize};
 use worker::*;
 
+/// Error Parameters used by the Leader to report errors caused by peers.
+#[derive(Deserialize, Serialize)]
+pub(crate) enum PeerErr {
+    ErrBatchOverlap,
+}
```
Comment (cjpatton) [inline daphne_worker/src/lib.rs:21]: I think I see what you're getting at here: There may be other reasons for a DO to signal an abort, and this would be the enum that defines them. However I think trying to be general here is cumbersome and makes internal APIs a little awkward. My preference would be to do away with this enum here and modify the enum in aggregate_store.rs as follows:

```rust
#[derive(Deserialize, Serialize)]
pub(crate) enum AggregateStoreResult {
    Ok(DapAggregateShare),
    ErrBatchOverlap,
}
```

This saves the caller of the AggregateStore DO from having to handle undefined behavior (e.g., suppose AggregateStore returned an error variant other than ErrBatchOverlap).

File: daphne_worker_test/helper.env [hunk 20]
```diff
@@ -16,6 +16,9 @@ DAP_BUCKET_KEY = 'f79c352056982bae1737e34bdac24d63'
 # Number of buckets.
 DAP_BUCKET_COUNT = 2
 
+# Global configurations.
+GLOB_CONFIG = '{"max_window_size": 100,"min_duration_start": 432000,"max_duration_end": 18000}'
```
Comment (cjpatton) [inline daphne_worker_test/helper.env:20]: Here and below
```suggestion
GLOBAL_CONFIG = '{"max_window_size": 100,"min_duration_start": 432000,"max_duration_end": 18000}'
```

File: daphne_worker_test/tests/e2e.rs [hunk 21]
```diff
@@ -537,3 +540,94 @@ async fn e2e_leader_collect_abort_invalid_batch_interval() {
     )
     .await;
 }
+
+#[tokio::test]
+#[cfg_attr(not(feature = "test_e2e"), ignore)]
+async fn e2e_leader_collect_abort_overlapping_batch_interval() {
+    let t = TestRunner::default().await;
+    let batch_interval = t.batch_interval();
+
```
Comment (cjpatton) [inline daphne_worker_test/tests/e2e.rs:549]: nit: Remove this newline

File: daphne_worker_test/tests/e2e.rs [hunk 22]
```diff
@@ -537,3 +540,94 @@ async fn e2e_leader_collect_abort_invalid_batch_interval() {
...
+    assert_eq!(
+        agg_telem.reports_aggregated, t.min_batch_size,
+        "reports aggregated"
+    );
+    assert_eq!(
+        agg_telem.reports_collected, t.min_batch_size,
+        "reports collected"
+    );
+
+    // Send a collect request that requests results from batches that overlaps with
```
Comment (cjpatton) [inline daphne_worker_test/tests/e2e.rs:603]:
```suggestion
    // Send a collect request that overlaps with
```

File: daphne_worker_test/tests/e2e.rs [hunk 23]
```diff
@@ -537,3 +540,94 @@ async fn e2e_leader_collect_abort_invalid_batch_interval() {
...
+        .internal_process(
+            &client,
+            &InternalAggregateInfo {
+                max_buckets: 100, // Needs to be sufficiently large to touch each bucket.
+                max_reports: 100,
+            },
+        )
+        .await
+        .unwrap_err();
+    assert_matches!(err, DapAbort::BatchOverlap);
```
Comment (cjpatton) [inline daphne_worker_test/tests/e2e.rs:632]: Hmm, this is odd. I would expect `POST /internal/process` to succeed (although no reports would be aggregated). Conversely, I would expect the second `POST /collect`above to fail. In fact, this is what the spec mandates.

File: daphne_worker_test/tests/e2e.rs [hunk 24]
```diff
@@ -537,3 +540,94 @@ async fn e2e_leader_collect_abort_invalid_batch_interval() {
...
+        task_id: t.task_id.clone(),
+        batch_interval: Interval {
+            start: batch_interval.start,
+            duration: batch_interval.duration * 2,
+        },
+        agg_param: Vec::new(),
+    };
+    let _collect_uri = t
+        .leader_post_collect(&client, collect_req.get_encoded())
+        .await;
```
Comment (cjpatton) [inline daphne_worker_test/tests/e2e.rs:619]: Why doesn't the test fail here?

**Review by cjpatton — CHANGES_REQUESTED** [review]

File: daphne/src/lib.rs (hunk 12 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs]: Completely forgot about this! Thanks for pointing it out.

File: daphne/src/lib.rs (hunk 13 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs]: I've changed the naming of this and the other variables in this struct.

File: daphne/src/lib.rs (hunk 14 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs]: See above.

File: daphne/src/lib.rs (hunk 15 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs]: This is a number of seconds.

File: daphne/src/lib.rs (hunk 16 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs]: Didn't know this :)
Updated Glob to Global

File: daphne_worker/src/config.rs (hunk 17 above)
Comment (nakatsuka-y) [reply daphne_worker/src/config.rs]: Ditto!

File: daphne/src/lib.rs (hunk 18 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs]: Changed `PeerError` to `Abort`.

File: daphne_worker/src/lib.rs (hunk 19 above)
Comment (nakatsuka-y) [reply daphne_worker/src/lib.rs]: I thought it would be nice to have a separate enum, but your point makes sense.
I've updated the enum.

File: daphne_worker_test/helper.env (hunk 20 above)
Comment (nakatsuka-y) [reply daphne_worker_test/helper.env]: Thanks :)

File: daphne_worker_test/tests/e2e.rs (hunk 21 above)
Comment (nakatsuka-y) [reply daphne_worker_test/tests/e2e.rs]: Sounds good.

File: daphne_worker_test/tests/e2e.rs (hunk 22 above)
Comment (nakatsuka-y) [reply daphne_worker_test/tests/e2e.rs]: Grammar issues :(

File: daphne_worker_test/tests/e2e.rs (hunk 23 above)
Comment (nakatsuka-y) [reply daphne_worker_test/tests/e2e.rs]: I've updated the code so that the collect req gets checked when it gets posted.

File: daphne_worker_test/tests/e2e.rs (hunk 24 above)
Comment (nakatsuka-y) [reply daphne_worker_test/tests/e2e.rs]: It should now.

File: daphne/src/lib.rs [hunk 25]
```diff
@@ -225,6 +236,25 @@ pub struct ProblemDetails {
     pub detail: Option<String>,
 }
 
+/// Global DAP parameters common across tasks.
+//
+// TODO spec: `max_window_size`, `min_duration_start`, and `max_duration_end`
```
Comment (cjpatton) [inline daphne/src/lib.rs:241]: These need to be renamed to match struct.

File: daphne/src/lib.rs [hunk 26]
```diff
@@ -225,6 +236,25 @@ pub struct ProblemDetails {
...
+//            should be reflected to the DAP spec.
+#[derive(Deserialize, Serialize)]
+pub struct DapGlobConfig {
+    /// Maximum window size permitted in CollectReq.
+    /// Prevents Collectors from requesting wide range or reports.
+    pub max_batch_duration: u64,
+
+    /// Lower bound of an acceptable batch window in CollectReq.
+    /// Batch intervals must start after `min_batch_interval_start`.
+    pub min_batch_interval_start: u64,
```
Comment (cjpatton) [inline daphne/src/lib.rs:251]: Is this a fixed point in time, or a duration between the current time and the min batch interval start? I think it should be the latter, otherwise we'd have to update this periodically.

File: daphne_worker_test/tests/test_runner.rs [hunk 27]
```diff
@@ -354,21 +388,25 @@ impl TestRunner {
         &self,
         client: &reqwest::Client,
         agg_info: &InternalAggregateInfo,
-    ) -> DapLeaderProcessTelemetry {
+    ) -> Result<DapLeaderProcessTelemetry, DapAbort> {
```
Comment (cjpatton) [inline daphne_worker_test/tests/test_runner.rs:391]: Let's revert this change, since we no longer need and it's causing lots of clutter in this PR.

**Review by cjpatton — COMMENTED** [review]
Alright this is heading in a good direction 👍 

Let's see if we can refactor so that check for overlapping batches can be moved into the `DapLeader` and `DapAggregator`. Here's what I suggest we do:

* Add a method to the `DapAggregator` trait for checking whether a given batch interval would overlap with a previous request.
* Modify `DapLeader::http_post_collect()` to call the new method
* Modify `DapHelper::http_post_aggregate_share()` to call the new method.

File: daphne/src/lib.rs (hunk 25 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs]: Thanks for pointing this out, it seemed that there were other locations that used the legacy variable name, so I've updated those in addition to this.

File: daphne/src/lib.rs (hunk 26 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs:251]: `min_batch_interval_start` and `max_batch_interval_end` are duration between the current time and the min/max batch interval start/end.
Do you have a better name for this variable?

File: daphne_worker_test/tests/test_runner.rs (hunk 27 above)
Comment (nakatsuka-y) [reply daphne_worker_test/tests/test_runner.rs]: Sounds good, I've done this.

File: daphne/src/lib.rs [hunk 28]
```diff
@@ -225,6 +236,25 @@ pub struct ProblemDetails {
     pub detail: Option<String>,
 }
 
+/// Global DAP parameters common across tasks.
+//
+// TODO spec: `max_batch_duration`, `min_batch_interval_start`,
+// and `max_batch_interval_end` should be reflected to the DAP spec.
```
Comment (cjpatton) [inline daphne/src/lib.rs:242]: Hmm, arguably this is not needed. Would you mind bringing this up in an issue? https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/new I'm wondering if anyone has thought about this yet.

File: daphne/src/lib.rs [hunk 29]
```diff
@@ -225,6 +236,25 @@ pub struct ProblemDetails {
...
+#[derive(Deserialize, Serialize)]
+pub struct DapGlobConfig {
+    /// Maximum window size permitted in CollectReq.
+    /// Prevents Collectors from requesting wide range or reports.
+    pub max_batch_duration: u64,
+
+    /// Lower bound of an acceptable batch window in CollectReq.
```
Comment (cjpatton) [inline daphne/src/lib.rs:249]: Let's avoid overloading the term "batch window", which has a different definition in the spec.

```suggestion
    /// Lower bound of an acceptable batch interval for collect requests.
```

File: daphne/src/lib.rs (hunk 26 above)
Comment (cjpatton) [reply daphne/src/lib.rs:251]: These names are fine.

File: daphne/src/roles_test.rs [hunk 30]
```diff
@@ -559,16 +690,157 @@ async fn poll_collect_job_test_results() {
...
+        payload: collector_collect_req.get_encoded(),
+        url: task_config.helper_url.join("/collect").unwrap(),
+        sender_auth: Some(BearerToken::from(COLLECTOR_BEARER_TOKEN.to_string())),
+    };
+
+    // Leader: Handle the CollectReq received from Collector.
+    let err = leader.http_post_collect(&req).await.unwrap_err();
+
+    // Fails because the requested batch interval is too large.
+    assert_matches!(err, DapAbort::InvalidBatchInterval);
```
Comment (cjpatton) [inline daphne/src/roles_test.rs:724]: The "invalidBatchInterval" has a very specific semantics in DAP-01, this is technically non-compliant. We should consider putting this in the spec, but in the mean time I think we need to handle this explicitly. Can we return `DapAbort::BadRequest` with a string that signifies why we're rejecting the report?

File: daphne_worker/src/dap.rs [hunk 31]
```diff
@@ -297,12 +356,27 @@ impl<D> DapLeader<BearerToken> for DaphneWorkerConfig<D> {
             .get_task_config_for(&collect_req.task_id)
             .ok_or_else(|| DapError::fatal(INT_ERR_UNRECOGNIZED_TASK))?;
```
Comment (cjpatton) [inline daphne_worker/src/dap.rs:358]: Sampling the clock twice can lead to strange errors. Sample it once, then use it twice
```rust
let now = now();
if some_condition_that_depends_on_the_current_time(now) {
 ....
}
```

File: daphne_worker/src/dap.rs [hunk 32]
```diff
@@ -297,12 +356,27 @@ impl<D> DapLeader<BearerToken> for DaphneWorkerConfig<D> {
...
 
+        if collect_req.batch_interval.duration / task_config.min_batch_duration
+            > self.global_config.max_batch_duration
+            || now().abs_diff(collect_req.batch_interval.start)
+                > self.global_config.min_batch_interval_start
+            || now().abs_diff(collect_req.batch_interval.end())
+                > self.global_config.max_batch_interval_end
+        {
+            return Err(DapError::Abort(DapAbort::InvalidBatchInterval));
```
Comment (cjpatton) [inline daphne_worker/src/dap.rs:366]: Just as for `MockAggregator`, this should return something like `DapAbort::BadRequest("batch interval ")`. You might also consider leaving a note here that we might consider addressing in the spec.

File: daphne_worker/src/durable/leader_col_job_queue.rs [hunk 33]
```diff
@@ -24,10 +24,10 @@ pub(crate) const DURABLE_LEADER_COL_JOB_QUEUE_FINISH: &str =
...
     "/internal/do/leader_col_job_queue/get_result";
 
-#[derive(Debug, Deserialize, Serialize)]
-pub(crate) struct CollectionJobResult {
-    pub(crate) collect_id: Id,
-    pub(crate) collect_resp: CollectResp,
+#[derive(Deserialize, Serialize)]
+pub(crate) enum CollectionJobQueueResult {
+    Ok(Id),
+    ErrBatchOverlap,
```
Comment (cjpatton) [inline daphne_worker/src/durable/leader_col_job_queue.rs:30]: This is no longer returned. Do we still need CollectionJobQueueResult? It seem slike this can be revereted

**Review by cjpatton — COMMENTED** [review]
Nice! Just a bit of clean up.

File: daphne/src/lib.rs (hunk 28 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs]: Since this is a Daphne-specific implementation detail, I've swapped the TODO with a note regarding whether to consider putting this into the spec or not.

File: daphne/src/lib.rs (hunk 29 above)
Comment (nakatsuka-y) [reply daphne/src/lib.rs]: Good point. I've also updated the comment so that it describes what these variables are in more detail.

File: daphne/src/roles_test.rs (hunk 30 above)
Comment (nakatsuka-y) [reply daphne/src/roles_test.rs]: I've used BadRequest and used the string to explicitly report about the reason.

File: daphne_worker/src/dap.rs (hunk 31 above)
Comment (nakatsuka-y) [reply daphne_worker/src/dap.rs:358]: I totally missed this. Thanks for pointing this out.

File: daphne_worker/src/dap.rs (hunk 32 above)
Comment (nakatsuka-y) [reply daphne_worker/src/dap.rs]: Changed!

File: daphne_worker/src/durable/leader_col_job_queue.rs (hunk 33 above)
Comment (nakatsuka-y) [reply daphne_worker/src/durable/leader_col_job_queue.rs]: Great point, I've removed this.

**Review by cjpatton — COMMENTED** [review]
LGTM, go ahead and squash

**Review by cjpatton — APPROVED** [review]
🦆
